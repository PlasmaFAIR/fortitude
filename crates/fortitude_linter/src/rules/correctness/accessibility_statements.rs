use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for missing `private` or `public` accessibility statements in modules
///
/// ## Why is this bad?
/// The `private` statement makes all entities (variables, types, procedures)
/// private by default, requiring an explicit `public` attribute to make them
/// available. As well as improving encapsulation between modules, this also
/// makes it possible to detect unused entities.
///
/// A `public` statement in a module does not change the default behaviour,
/// and therefore all entities will be available from outside the module
/// unless they are individually given a `private` attribute. This brings
/// all of the same downsides as the default behaviour, but an explicit
/// `public` statement makes it clear that the programmer is choosing
/// this behaviour intentionally.  
#[derive(ViolationMetadata)]
pub(crate) struct MissingAccessibilityStatement {
    name: String,
}

impl Violation for MissingAccessibilityStatement {
    #[derive_message_formats]
    fn message(&self) -> String {
        format!(
            "module '{}' missing default accessibility statement",
            self.name
        )
    }
}

impl AstRule for MissingAccessibilityStatement {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let module = node.parent()?;

        let bare_private_statement = match module.child_with_name("private_statement") {
            Some(statement) => statement.named_child(0).is_none(),
            None => false,
        };

        let bare_public_statement = match module.child_with_name("public_statement") {
            Some(statement) => statement.named_child(0).is_none(),
            None => false,
        };

        // No statement whatsoever
        if !bare_private_statement && !bare_public_statement {
            let name = node.named_child(0)?.to_text(src.source_text())?.to_string();
            return some_vec![Diagnostic::from_node(
                MissingAccessibilityStatement { name },
                node
            )];
        }

        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["module_statement"]
    }
}

/// ## What it does
/// Checks if the default accessibility in modules is set to `public`
///
/// ## Why is this bad?
/// The `public` statement makes all entities (variables, types, procedures)
/// public by default. This decreases encapsulation and makes it more likely to
/// accidentally expose more than necessary. Public accessibility also makes
/// it harder to detect unused entities, which can often be indicative of
/// errors within the code.
#[derive(ViolationMetadata)]
pub(crate) struct DefaultPublicAccessibility {
    name: String,
}

impl Violation for DefaultPublicAccessibility {
    #[derive_message_formats]
    fn message(&self) -> String {
        format!("module '{}' has default `public` accessibility", self.name)
    }
}

impl AstRule for DefaultPublicAccessibility {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // Bare `public` statement`
        if node.named_child(0).is_none() {
            let module = node.parent()?;
            let statement = module.child_with_name("module_statement")?;
            let name = statement
                .child_with_name("name")?
                .to_text(src.source_text())?
                .to_string();
            return some_vec![Diagnostic::from_node(
                DefaultPublicAccessibility { name },
                node
            )];
        }

        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["public_statement"]
    }
}
