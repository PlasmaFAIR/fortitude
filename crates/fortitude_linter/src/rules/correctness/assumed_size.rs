use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use itertools::Itertools;
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What does it do?
/// Checks for assumed size variables
///
/// ## Why is this bad?
/// Assumed size dummy arguments declared with a star `*` as the size should be
/// avoided. There are several downsides to assumed size, the main one being
/// that the compiler is not able to determine the array bounds, so it is not
/// possible to check for array overruns or to use the array in whole-array
/// expressions.
///
/// Instead, prefer assumed shape arguments, as the compiler is able to keep track of
/// the upper bounds automatically, and pass this information under the hood. It also
/// allows use of whole-array expressions, such as `a = b + c`, where `a, b, c` are
/// all arrays of the same shape.
///
/// Instead of:
///
/// ```f90
/// subroutine process_array(array)
///     integer, dimension(*), intent(in) :: array
///     ...
/// ```
///
/// use:
///
/// ```f90
/// subroutine process_array(array)
///     integer, dimension(:), intent(in) :: array
///     ...
/// ```
///
/// Note that this doesn't apply to `character` types, where `character(len=*)` is
/// actually the most appropriate specification for `intent(in)` arguments! This is
/// because `character(len=:)` must be either a `pointer` or `allocatable`.
#[derive(ViolationMetadata)]
pub(crate) struct AssumedSize {
    name: String,
}

impl Violation for AssumedSize {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("'{name}' has assumed size")
    }
}
impl AstRule for AssumedSize {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = src.source_text();
        let declaration = node
            .ancestors()
            .find(|parent| parent.kind() == "variable_declaration")?;

        // Deal with `character([len=]*)` elsewhere
        if let Some(dtype) = declaration.parse_intrinsic_type() {
            let is_character = dtype.to_lowercase() == "character";
            let is_kind = node.ancestors().any(|parent| parent.kind() == "kind");
            if is_character && is_kind {
                return None;
            }
        }

        // Assumed size ok for parameters
        if declaration
            .children_by_field_name("attribute", &mut declaration.walk())
            .filter_map(|attr| attr.to_text(src))
            .any(|attr_name| attr_name.to_lowercase() == "parameter")
        {
            return None;
        }

        // Are we looking at something declared like `array(*)`?
        if let Some(sized_decl) = node
            .ancestors()
            .find(|parent| parent.kind() == "sized_declarator")
        {
            let identifier = sized_decl.child_with_name("identifier")?;
            let name = identifier.to_text(src)?.to_string();
            return some_vec![Diagnostic::from_node(Self { name }, node)];
        }

        // Collect things that look like `dimension(*)` -- this
        // applies to all identifiers on this line
        let all_decls = declaration
            .children_by_field_name("declarator", &mut declaration.walk())
            .filter_map(|declarator| {
                let identifier = match declarator.kind() {
                    "identifier" => Some(declarator),
                    "sized_declarator" => declarator.child_with_name("identifier"),
                    _ => None,
                }?;
                identifier.to_text(src)
            })
            .map(|name| name.to_string())
            .map(|name| Diagnostic::from_node(Self { name }, node))
            .collect_vec();

        Some(all_decls)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["assumed_size"]
    }
}

/// ## What does it do?
/// Checks `character` dummy arguments have `intent(in)` only
///
/// ## Why is this bad?
/// Character dummy arguments with an assumed size should only have `intent(in)`, as
/// this can cause data loss with `intent([in]out)`. For example:
///
/// ```f90
/// program example
///   character(len=3) :: short_text
///   call set_text(short_text)
///   print*, short_text
/// contains
///   subroutine set_text(text)
///     character(*), intent(out) :: text
///     text = "longer than 3 characters"
///   end subroutine set_text
/// end program
/// ```
///
/// Here, `short_text` will only contain the truncated "lon".
///
/// To handle dynamically setting `character` sizes, use `allocatable` instead:
///
/// ```f90
/// program example
///   character(len=:), allocatable :: allocatable_text
///   call set_text(allocatable_text)
///   print*, allocatable_text
/// contains
///   subroutine set_text(text)
///     character(len=:), allocatable, intent(out) :: text
///     text = "longer than 3 characters"
///   end subroutine set_text
/// end program
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct AssumedSizeCharacterIntent {
    name: String,
}

impl Violation for AssumedSizeCharacterIntent {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("character '{name}' has assumed size but does not have `intent(in)`")
    }
}
impl AstRule for AssumedSizeCharacterIntent {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = src.source_text();
        // TODO: This warning will also catch:
        // - non-dummy arguments -- these are always invalid, should be a separate warning?

        let declaration = node
            .ancestors()
            .find(|parent| parent.kind() == "variable_declaration")?;

        // Only applies to `character`
        if declaration.parse_intrinsic_type()?.to_lowercase() != "character" {
            return None;
        }

        // Handle `character*(*)` elsewhere -- note this just skips emitting a warning
        // for the first `*`, we'll still get one for the second `*`, but this is desired
        if let Some(sibling) = node.next_named_sibling() {
            if sibling.kind() == "assumed_size" {
                return None;
            }
        }

        let attrs_as_text = declaration
            .children_by_field_name("attribute", &mut declaration.walk())
            .filter_map(|attr| attr.to_text(src))
            .map(|attr| attr.to_lowercase())
            .collect_vec();

        // Assumed size ok for parameters
        if attrs_as_text.iter().any(|attr| attr == "parameter") {
            return None;
        }

        // Ok for `intent(in)` only
        if let Some(intent) = attrs_as_text.iter().find(|attr| attr.starts_with("intent")) {
            let intent = intent.split_whitespace().collect_vec().join("");
            if intent == "intent(in)" {
                return None;
            }
        }

        // Collect all declarations on this line
        let all_decls = declaration
            .children_by_field_name("declarator", &mut declaration.walk())
            .filter_map(|declarator| {
                let identifier = match declarator.kind() {
                    "identifier" => Some(declarator),
                    "sized_declarator" => declarator.child_with_name("identifier"),
                    _ => None,
                }?;
                identifier.to_text(src)
            })
            .map(|name| name.to_string())
            .map(|name| Diagnostic::from_node(Self { name }, node))
            .collect_vec();

        Some(all_decls)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["assumed_size"]
    }
}
