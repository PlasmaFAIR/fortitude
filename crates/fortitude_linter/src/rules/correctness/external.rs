use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

use crate::{ast::FortitudeNode, settings::Settings, AstRule, FromAstNode};

/// ## What does it do?
/// Checks for procedures declared with just `external`
///
/// ## Why is this bad?
/// Compilers are unable to check external procedures without an explicit
/// interface for errors such as wrong number or type of arguments.
///
/// If the procedure is in your project, put it in a module (see
/// `external-function`), or write an explicit interface.
#[derive(ViolationMetadata)]
pub(crate) struct ExternalProcedure {
    name: String,
}

impl Violation for ExternalProcedure {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("'{name}' declared as `external`")
    }

    fn fix_title(&self) -> Option<String> {
        Some(format!("Write an explicit interface"))
    }
}

impl AstRule for ExternalProcedure {
    fn check(_settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
        if node
            .child_with_name("type_qualifier")?
            .to_text(source.source_text())?
            .to_lowercase()
            != "external"
        {
            return None;
        }

        let name = node
            .child_by_field_name("declarator")?
            .to_text(source.source_text())?
            .to_string();
        some_vec!(Diagnostic::from_node(Self { name }, node))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["variable_modification"]
    }
}

/// ## What it does
/// Checks for any functions and subroutines not defined within modules (or one
/// of a few acceptable alternatives).
///
/// ## Why is this bad?
/// Functions and subroutines should be contained within (sub)modules or programs.
/// Fortran compilers are unable to perform type checks and conversions on functions
/// defined outside of these scopes, and this is a common source of bugs.
#[derive(ViolationMetadata)]
pub(crate) struct ProcedureNotInModule {
    procedure: String,
}

impl Violation for ProcedureNotInModule {
    #[derive_message_formats]
    fn message(&self) -> String {
        let ProcedureNotInModule { procedure } = self;
        format!("{procedure} not contained within (sub)module or program")
    }
}

impl AstRule for ProcedureNotInModule {
    fn check(_settings: &Settings, node: &Node, _src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if node.parent()?.kind() == "translation_unit" {
            let procedure_stmt = node.child(0)?;
            let procedure = node.kind().to_string();
            return some_vec![Diagnostic::from_node(
                ProcedureNotInModule { procedure },
                &procedure_stmt
            )];
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["function", "subroutine"]
    }
}
