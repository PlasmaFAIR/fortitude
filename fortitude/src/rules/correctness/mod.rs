pub mod accessibility_statements;
pub mod assumed_size;
pub mod derived_default_init;
pub mod exit_labels;
pub mod external;
pub mod implicit_kinds;
pub mod implicit_typing;
pub mod init_decls;
pub mod intent;
pub mod kind_suffixes;
pub mod magic_numbers;
pub mod missing_io_specifier;
pub mod select_default;
pub mod trailing_backslash;
pub mod use_statements;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::rules::correctness::exit_labels;
    use crate::settings::{CheckSettings, Settings};
    use crate::test::test_path;

    #[test_case(Rule::ImplicitTyping, Path::new("C001.f90"))]
    #[test_case(Rule::InterfaceImplicitTyping, Path::new("C002.f90"))]
    #[test_case(Rule::ImplicitExternalProcedures, Path::new("C003.f90"))]
    #[test_case(Rule::MissingDefaultCase, Path::new("C011.f90"))]
    #[test_case(Rule::NoRealSuffix, Path::new("C021.f90"))]
    #[test_case(Rule::ImplicitRealKind, Path::new("C022.f90"))]
    #[test_case(Rule::MagicNumberInArraySize, Path::new("C031.f90"))]
    #[test_case(Rule::MagicIoUnit, Path::new("C032.f90"))]
    #[test_case(Rule::MissingActionSpecifier, Path::new("C041.f90"))]
    #[test_case(Rule::TrailingBackslash, Path::new("C051.F90"))]
    #[test_case(Rule::MissingIntent, Path::new("C061.f90"))]
    #[test_case(Rule::AssumedSize, Path::new("C071.f90"))]
    #[test_case(Rule::AssumedSizeCharacterIntent, Path::new("C072.f90"))]
    #[test_case(Rule::InitialisationInDeclaration, Path::new("C081.f90"))]
    #[test_case(Rule::PointerInitialisationInDeclaration, Path::new("C082.f90"))]
    #[test_case(Rule::ExternalProcedure, Path::new("C091.f90"))]
    #[test_case(Rule::ProcedureNotInModule, Path::new("C092.f90"))]
    #[test_case(Rule::MissingDefaultPointerInitalisation, Path::new("C101.f90"))]
    #[test_case(Rule::UseAll, Path::new("C121.f90"))]
    #[test_case(Rule::MissingIntrinsic, Path::new("C122.f90"))]
    #[test_case(Rule::MissingAccessibilityStatement, Path::new("C131.f90"))]
    #[test_case(Rule::DefaultPublicAccessibility, Path::new("C132.f90"))]
    #[test_case(Rule::MissingExitOrCycleLabel, Path::new("C141.f90"))]
    #[test_case(Rule::ExitOrCycleInUnlabelledLoop, Path::new("C142.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::ImplicitTyping, Path::new("C001_ok.f90"))]
    #[test_case(Rule::InterfaceImplicitTyping, Path::new("C002_ok.f90"))]
    fn rules_pass(rule_code: Rule, path: &Path) -> Result<()> {
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        assert!(
            diagnostics.is_empty(),
            "Test source has no warnings, but some were raised:\n{diagnostics}"
        );
        Ok(())
    }

    #[test]
    fn warn_c142_on_nested_loops_only() -> Result<()> {
        let rule_code = Rule::ExitOrCycleInUnlabelledLoop;
        let path = Path::new("C142.f90");
        let snapshot = format!(
            "{}_{}_nested_loops_only",
            rule_code.as_ref(),
            path.to_string_lossy()
        );
        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                exit_unlabelled_loops: exit_labels::settings::Settings {
                    allow_unnested_loops: true,
                },
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
