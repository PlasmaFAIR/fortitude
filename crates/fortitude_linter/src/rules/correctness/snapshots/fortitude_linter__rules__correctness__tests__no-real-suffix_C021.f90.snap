---
source: fortitude/src/rules/correctness/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/correctness/C021.f90:13:30: C021 real literal 1.23456 missing kind suffix
   |
11 |   real(sp), parameter :: e = 1.0e10 ! Okay: No loss of precision, e exponent
12 |   real(sp), parameter :: f = -2E10 ! Okay: No loss of precision, E exponent
13 |   real(sp), parameter :: g = 1.23456 ! Bad: Loss of precision
   |                              ^^^^^^^ C021
14 |   real(sp), parameter :: h = 1.23456e1 ! Bad: Loss of precision, e exponent
15 |   real(sp), parameter :: i = 1.23456E-1 ! Bad: Loss of precision, E exponent
   |

./resources/test/fixtures/correctness/C021.f90:14:30: C021 real literal 1.23456e1 missing kind suffix
   |
12 |   real(sp), parameter :: f = -2E10 ! Okay: No loss of precision, E exponent
13 |   real(sp), parameter :: g = 1.23456 ! Bad: Loss of precision
14 |   real(sp), parameter :: h = 1.23456e1 ! Bad: Loss of precision, e exponent
   |                              ^^^^^^^^^ C021
15 |   real(sp), parameter :: i = 1.23456E-1 ! Bad: Loss of precision, E exponent
16 |   real(dp), parameter :: j = -1.23456_dp ! Okay: Kind suffix
   |

./resources/test/fixtures/correctness/C021.f90:15:30: C021 real literal 1.23456E-1 missing kind suffix
   |
13 |   real(sp), parameter :: g = 1.23456 ! Bad: Loss of precision
14 |   real(sp), parameter :: h = 1.23456e1 ! Bad: Loss of precision, e exponent
15 |   real(sp), parameter :: i = 1.23456E-1 ! Bad: Loss of precision, E exponent
   |                              ^^^^^^^^^^ C021
16 |   real(dp), parameter :: j = -1.23456_dp ! Okay: Kind suffix
17 |   real(sp), parameter :: k = 1.23456_sp ! Okay: Loss of precision, but we're explicit
   |

./resources/test/fixtures/correctness/C021.f90:20:30: C021 real literal 2e39 missing kind suffix
   |
18 |   real(dp), parameter :: l = 1.23456d1 ! Okay: Ignore d exponent
19 |   real(dp), parameter :: m = 1.23456D-3 ! Okay: Ignore D exponent
20 |   real(dp), parameter :: n = 2e39 ! Bad: Magnitude doesn't fit in single precision
   |                              ^^^^ C021
21 |   real(dp), parameter :: o = -(+(-(((3.141592654))))) ! Bad: Loss of precision, horrible declaration
   |

./resources/test/fixtures/correctness/C021.f90:21:38: C021 real literal 3.141592654 missing kind suffix
   |
19 |   real(dp), parameter :: m = 1.23456D-3 ! Okay: Ignore D exponent
20 |   real(dp), parameter :: n = 2e39 ! Bad: Magnitude doesn't fit in single precision
21 |   real(dp), parameter :: o = -(+(-(((3.141592654))))) ! Bad: Loss of precision, horrible declaration
   |                                      ^^^^^^^^^^^ C021
22 |   
23 |   complex(dp), parameter :: ca = (0.0, 0.0) ! Okay: No loss of precision
   |

./resources/test/fixtures/correctness/C021.f90:28:42: C021 real literal 4.0E-5 missing kind suffix
   |
26 |   complex(dp), parameter :: cd = (0.000244140625, 0.0) ! Okay: No loss of precision (=2^-12)
27 |   complex(sp), parameter :: ce = (1.0e10, 2.0e5) ! Okay: No loss of precision, e exponent
28 |   complex(sp), parameter :: cf = (2E10, -4.0E-5) ! Bad: Loss of precision, negative E exponent
   |                                          ^^^^^^ C021
29 |   complex(sp), parameter :: cg = (1.23456, -0.0) ! Bad: Loss of precision
30 |   complex(sp), parameter :: ch = (0.0, 1.23456e1) ! Bad: Loss of precision, e exponent
   |

./resources/test/fixtures/correctness/C021.f90:29:35: C021 real literal 1.23456 missing kind suffix
   |
27 |   complex(sp), parameter :: ce = (1.0e10, 2.0e5) ! Okay: No loss of precision, e exponent
28 |   complex(sp), parameter :: cf = (2E10, -4.0E-5) ! Bad: Loss of precision, negative E exponent
29 |   complex(sp), parameter :: cg = (1.23456, -0.0) ! Bad: Loss of precision
   |                                   ^^^^^^^ C021
30 |   complex(sp), parameter :: ch = (0.0, 1.23456e1) ! Bad: Loss of precision, e exponent
31 |   complex(sp), parameter :: ci = (1.23456E1, 0.0) ! Bad: Loss of precision, E exponent
   |

./resources/test/fixtures/correctness/C021.f90:30:40: C021 real literal 1.23456e1 missing kind suffix
   |
28 |   complex(sp), parameter :: cf = (2E10, -4.0E-5) ! Bad: Loss of precision, negative E exponent
29 |   complex(sp), parameter :: cg = (1.23456, -0.0) ! Bad: Loss of precision
30 |   complex(sp), parameter :: ch = (0.0, 1.23456e1) ! Bad: Loss of precision, e exponent
   |                                        ^^^^^^^^^ C021
31 |   complex(sp), parameter :: ci = (1.23456E1, 0.0) ! Bad: Loss of precision, E exponent
32 |   complex(dp), parameter :: cj = (-1.23456_dp, 0.2_dp) ! Okay: Kind suffix
   |

./resources/test/fixtures/correctness/C021.f90:31:35: C021 real literal 1.23456E1 missing kind suffix
   |
29 |   complex(sp), parameter :: cg = (1.23456, -0.0) ! Bad: Loss of precision
30 |   complex(sp), parameter :: ch = (0.0, 1.23456e1) ! Bad: Loss of precision, e exponent
31 |   complex(sp), parameter :: ci = (1.23456E1, 0.0) ! Bad: Loss of precision, E exponent
   |                                   ^^^^^^^^^ C021
32 |   complex(dp), parameter :: cj = (-1.23456_dp, 0.2_dp) ! Okay: Kind suffix
33 |   complex(sp), parameter :: ck = (1.23456_sp, 3.14159265_sp)! Okay: Loss of precision, but we're explicit
   |

./resources/test/fixtures/correctness/C021.f90:36:35: C021 real literal 2e39 missing kind suffix
   |
34 |   complex(dp), parameter :: cl = (1.23456d1, 37d2) ! Okay: Ignore d exponent
35 |   complex(dp), parameter :: cm = (1.23456D3, 37D2) ! Okay: Ignore D exponent
36 |   complex(dp), parameter :: cn = (2e39, 0e0) ! Bad: Magnitude doesn't fit in single precision
   |                                   ^^^^ C021
37 |   real(dp), parameter :: co = -(+(-(-3.141592654, +5.182647))) ! Bad: Loss of precision, horrible declaration
   |

./resources/test/fixtures/correctness/C021.f90:37:38: C021 real literal 3.141592654 missing kind suffix
   |
35 |   complex(dp), parameter :: cm = (1.23456D3, 37D2) ! Okay: Ignore D exponent
36 |   complex(dp), parameter :: cn = (2e39, 0e0) ! Bad: Magnitude doesn't fit in single precision
37 |   real(dp), parameter :: co = -(+(-(-3.141592654, +5.182647))) ! Bad: Loss of precision, horrible declaration
   |                                      ^^^^^^^^^^^ C021
38 |
39 |   real(dp) :: p, q, x, y, z
   |

./resources/test/fixtures/correctness/C021.f90:37:52: C021 real literal 5.182647 missing kind suffix
   |
35 |   complex(dp), parameter :: cm = (1.23456D3, 37D2) ! Okay: Ignore D exponent
36 |   complex(dp), parameter :: cn = (2e39, 0e0) ! Bad: Magnitude doesn't fit in single precision
37 |   real(dp), parameter :: co = -(+(-(-3.141592654, +5.182647))) ! Bad: Loss of precision, horrible declaration
   |                                                    ^^^^^^^^ C021
38 |
39 |   real(dp) :: p, q, x, y, z
   |

./resources/test/fixtures/correctness/C021.f90:42:12: C021 real literal 2.0 missing kind suffix
   |
40 |   complex(dp) :: cp, cq, cx, cy, cz
41 |
42 |   x = sqrt(2.0) ! Bad: Loss of precision
   |            ^^^ C021
43 |   y = real(1.0, kind=dp) ! Okay: Type cast with no loss of precision
44 |   z = real(1.0 + 1.0, kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |

./resources/test/fixtures/correctness/C021.f90:44:12: C021 real literal 1.0 missing kind suffix
   |
42 |   x = sqrt(2.0) ! Bad: Loss of precision
43 |   y = real(1.0, kind=dp) ! Okay: Type cast with no loss of precision
44 |   z = real(1.0 + 1.0, kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |            ^^^ C021
45 |   p = real(-5.0, kind=dp) ! Okay: Type cast with no loss of precision
46 |   q = real(1.23456, kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/correctness/C021.f90:44:18: C021 real literal 1.0 missing kind suffix
   |
42 |   x = sqrt(2.0) ! Bad: Loss of precision
43 |   y = real(1.0, kind=dp) ! Okay: Type cast with no loss of precision
44 |   z = real(1.0 + 1.0, kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |                  ^^^ C021
45 |   p = real(-5.0, kind=dp) ! Okay: Type cast with no loss of precision
46 |   q = real(1.23456, kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/correctness/C021.f90:46:12: C021 real literal 1.23456 missing kind suffix
   |
44 |   z = real(1.0 + 1.0, kind=dp) ! Bad: Type cast from expression, possible l.o.p
45 |   p = real(-5.0, kind=dp) ! Okay: Type cast with no loss of precision
46 |   q = real(1.23456, kind=dp) ! Bad: Type cast with loss of precision
   |            ^^^^^^^ C021
47 |   
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
   |

./resources/test/fixtures/correctness/C021.f90:48:14: C021 real literal 2.0 missing kind suffix
   |
46 |   q = real(1.23456, kind=dp) ! Bad: Type cast with loss of precision
47 |   
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
   |              ^^^ C021
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |

./resources/test/fixtures/correctness/C021.f90:48:19: C021 real literal 0.0 missing kind suffix
   |
46 |   q = real(1.23456, kind=dp) ! Bad: Type cast with loss of precision
47 |   
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
   |                   ^^^ C021
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |

./resources/test/fixtures/correctness/C021.f90:50:15: C021 real literal 1.0 missing kind suffix
   |
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |               ^^^ C021
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/correctness/C021.f90:50:20: C021 real literal 0.0 missing kind suffix
   |
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |                    ^^^ C021
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/correctness/C021.f90:50:28: C021 real literal 0.0 missing kind suffix
   |
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |                            ^^^ C021
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/correctness/C021.f90:50:33: C021 real literal 1.0 missing kind suffix
   |
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |                                 ^^^ C021
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/correctness/C021.f90:52:16: C021 real literal 1.23456 missing kind suffix
   |
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |                ^^^^^^^ C021
53 | end program test
   |

./resources/test/fixtures/correctness/C021.f90:52:25: C021 real literal 3.141292654 missing kind suffix
   |
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |                         ^^^^^^^^^^^ C021
53 | end program test
   |
